main.py

```diff
@@ -1510,42 +1513,87 @@ def show_hover(self, point, contents):
             max_width=800)
 
 
-class CompletionHandler(sublime_plugin.EventListener):
-    def __init__(self):
-        self.completions = []  # type: List[Tuple[str, str]]
-        self.refreshing = False
+class CompletionState(object):
+    IDLE = 0
+    REQUESTING = 1
+    APPLYING = 2
+    CANCELLING = 3
 
-    def on_query_completions(self, view, prefix, locations):
-        if not is_supported_view(view):
-            return None
 
-        if not self.refreshing:
-            client = client_for_view(view)
+class CompletionHandler(sublime_plugin.ViewEventListener):
+    def __init__(self, view):
+        self.view = view
+        self.initialized = False
+        self.enabled = False
+        self.trigger_chars = []  # type: List[str]
+        self.completions = []  # type: List[Tuple[str, str]]
+        self.state = CompletionState.IDLE
+        self.next_request = None
 
-            if not client:
-                return
+    @classmethod
+    def is_applicable(cls, settings):
+        syntax = settings.get('syntax')
+        return syntax and is_supported_syntax(syntax)
 
-            completionProvider = client.get_capability('completionProvider')
-            if not completionProvider:
+    def initialize(self):
+        self.initialized = True
+        client = client_for_view(self.view)
+        if client:
+            completionProvider = client.get_capability(
+                'completionProvider')
+            if completionProvider:
+                self.enabled = True
+                self.trigger_chars = completionProvider.get(
+                    'triggerCharacters') or []
                 return
 
-            autocomplete_triggers = completionProvider.get('triggerCharacters')
+    def is_after_trigger_character(self, location):
+        if location > 0:
+            prev_char = self.view.substr(
+                sublime.Region(location - 1, location))
+            return prev_char in self.trigger_chars
+
+    def on_query_completions(self, prefix, locations):
+        if not self.initialized:
+            self.initialize()
+
+        if self.enabled:
+            debug('enter at state {}'.format(self.state))
+            if self.state == CompletionState.IDLE:
+                self.do_request(prefix, locations)
+                self.completions = []  # type: List[Tuple[str, str]]
 
-            if locations[0] > 0:
-                self.completions = []
-                prev_char = view.substr(
-                    sublime.Region(locations[0] - 1, locations[0]))
-                if prev_char not in autocomplete_triggers:
-                    return None
+            elif self.state == CompletionState.REQUESTING or self.state == CompletionState.CANCELLING:
```

Comment:
How about: `elif self.state in (CompletionState.REQUESTING, CompletionState.CANCELLING)`

Reply:
Hmm. Seems quite easy to get the state machine stalled in CANCELLING state, once a language-server does not respond. From this point I'll never get any completion any more.

Reply:
Nice code suggestion. Why does the language server not respond?